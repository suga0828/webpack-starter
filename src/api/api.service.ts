import { User } from "./api.model";

const BASE_API = 'https://jsonplaceholder.typicode.com/users/';

export const irABuscarUsuarios = (): Promise<User[]> => fetch(BASE_API).then(resp => resp.json());

export const irABuscarUsuario = (id: string): Promise<User> => fetch(BASE_API + id).then(resp => resp.json());

export const crearUsuario = async (user: User): Promise<any> => {
  const config = {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(user)
  }
  const resp = await fetch(BASE_API, config);
  return await resp.json();
}

export const actualizarUsuario = async (user: Partial<User>): Promise<any> => {
  const config = {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ...user })
  }
  const response = await fetch(BASE_API + user.id, config);
  return await response.json();
}

export const eliminarUsuario = (id: number): Promise<any> => fetch(BASE_API + id, { method: 'DELETE' })
  .then(response => response.json());
