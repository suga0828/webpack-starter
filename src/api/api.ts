// const BASE_API = 'https://jsonplaceholder.typicode.com/users/';

// export const irABuscarUsuarios = () => fetch(BASE_API).then(resp => resp.json())

// export const irABuscarUsuario = () => fetch(BASE_API).then(resp => resp.json())

// export const crearUsuario = () => {
//   const config = {
//     method: 'POST',
//     headers: {
//       'Content-Type': 'application/json',
//     },
//     body: JSON.stringify({
//       name: 'toy name',
//       image: 'toy image',
//       likes: 0
//     })
//   }
//   return fetch(BASE_API, config).then(resp => resp.json())
// }

// export const actualizarUsuario = (toy) => {
//   const config = {
//     method: 'PATCH',
//     headers: {
//       'Content-Type': 'application/json',
//     },
//     body: JSON.stringify({
//       likes: toy.likes + 1
//     })
//   }
//   return fetch(BASE_API + toy.id, config).then(response => response.json())
// }

// export const eliminarUsuario = (id) => fetch(`${BASE_API}/${id}`, { method: 'DELETE' }).then(response => response.json())
