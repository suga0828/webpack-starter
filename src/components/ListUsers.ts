import { irABuscarUsuarios } from '../api/api.service';
import { User } from '../api/api.model';

export const imprimirUsiariosEnContenedor = async (contenedor: Element): Promise<void> => {
  try {
    const users: User[] = await irABuscarUsuarios();
    contenedor.innerHTML = '';

    users.forEach(user => {
      const liElement: HTMLLIElement = document.createElement('li');
      liElement.innerHTML = user.name;
      contenedor.insertAdjacentElement('beforeend', liElement);
    });
  }
  catch (err) {
    contenedor.innerHTML = 'Hubo un error: ' + err;
  }
}

export const Render = (): Element => {
  const idBoton = 'botonObtener'

  const template = `
    <h3>
      Lista de Usuarios
      <button id="${idBoton}">Obtener Usuarios</button>
    </h3>
    <ul id="listaUsuarios"></ul>
  `;

  const container = document.createElement('section');
  container.innerHTML = template;

  const botonObtenerUsuarios: HTMLButtonElement =  container.querySelector(`#${idBoton}`);
  const listContainer: HTMLUListElement = container.querySelector("#listaUsuarios");
  botonObtenerUsuarios.addEventListener('click', (e) => imprimirUsiariosEnContenedor(listContainer));

  return container;
};
