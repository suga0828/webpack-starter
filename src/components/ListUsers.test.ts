import * as Component from './ListUsers';

import * as API from '../api/api.service';
import { User } from '../api/api.model';

describe('ListUsers component', () => {
  describe('imprimirUsiariosEnContenedor method test suite', () => {
    it('should print users response inside container', async () => {
      jest.spyOn(API, 'irABuscarUsuarios')
        .mockImplementation((): Promise<User[]> => (
          new Promise(resolve => resolve([{ name: 'mock user' } as User]))
      ));

      const container: HTMLUListElement = document.createElement('ul');

      Component.imprimirUsiariosEnContenedor(container).then((e) => {
        expect(container.innerHTML).toBe('<li>mock user</li>');
      });
    });

    it('should print error response inside container', async () => {
      jest.spyOn(API, 'irABuscarUsuarios')
        .mockImplementation((): Promise<User[]> => (
          new Promise((_, reject) => reject('Error'))
      ));

      const container: HTMLUListElement = document.createElement('ul');

      Component.imprimirUsiariosEnContenedor(container).then((e) => {
        expect(container.innerHTML).toBe('Hubo un error: Error');
      });
    });
  });

  describe('ListUsers method test suite', () => {
    it('should return a template with a button with id botonObtener', () => {
      const template = Component.Render();

      expect(template.querySelector('#botonObtener')).not.toBeNull();
    });

    it('should return a template with an ul with id listaUsuarios', () => {
      const template = Component.Render();

      expect(template.querySelector('#listaUsuarios')).not.toBeNull();
    });

    describe('when make click on a button with id botonObtener', () => {
      it('should call imprimirUsiariosEnContenedor', () => {
        const spy = jest.spyOn(Component, 'imprimirUsiariosEnContenedor');
        const boton = Component.Render().querySelector('#botonObtener');

        boton.dispatchEvent(new Event('click'));
  
        expect(spy).toHaveBeenCalled();
      });
    })
  });
});

