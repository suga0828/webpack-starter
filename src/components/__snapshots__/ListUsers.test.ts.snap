// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListUsers component ListUsers method test suite should return ListUser component template 1`] = `
<section>
  
    
  <h3>
    
      Lista de Usuarios
      
    <button
      id="botonObtener"
    >
      Obtener Usuarios
    </button>
    
    
  </h3>
  
    
  <ul
    id="listaUsuarios"
  />
  
  
</section>
`;
